### PR TITLE
Replace git submodule with cmake fetch

### PR DESCRIPTION
**Pull request by cianciosa**
NOTE: This cannot be merged until the kronmult https://github.com/project-asgard/kronmult/pull/9 is merged since it requires changes from that pull request.

## Proposed changes

This change defines a macro function for registering external cmake based projects. Once an external project is registered, a developer may configure their build to point to other forks or other branches using the new  build setting

- `BUILD_REPO_<NAME>`
- `BUILD_TAG_<NAME>`

## What type(s) of changes does this code introduce?
_Put an `x` in the boxes that apply._

- [ ] Bugfix
- [ ] New feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Does this introduce a breaking change?

- [x] Yes
- [ ] No

## What systems has this change been tested on?
This has been tested with CUDA support on an ubuntu 18.04 install. Tests for

- kronmult-test
- time_advance-test-mpi
- time_advance-test
- kronmult_cuda-test

currently fail.

MacOS build still fail due to issues referenced in https://github.com/project-asgard/asgard/issues/387 in the adapt library.

## Checklist

- [x] this PR is up to date with current the current state of 'develop'
- [x] code added or changed in the PR has been clang-formatted
- [ ] this PR adds tests to cover any new code, or to catch a bug that is being fixed
- [ ] documentation has been added (if appropriate)

### CODE REVIEW

**Comment by cianciosa** [comment]
AWS tests fail at compile time due to

```
[ 29%] Building CXX object _deps/kronmult-build/CMakeFiles/kgemm_nn_test.dir/test_kgemm_nn_batched.cpp.o
--
801 | /codebuild/output/src458844159/src/github.com/project-asgard/asgard/src/device/kronmult_cuda.cpp:21:10: fatal error: kronmult1_xbatched.hpp: No such file or directory
802 | 21 \| #include "kronmult1_xbatched.hpp"
803 | \|          ^~~~~~~~~~~~~~~~~~~~~~~~
804 | compilation terminated.
```

The changes in https://github.com/project-asgard/kronmult/pull/9 are expected to resolve this.

File: CMakeLists.txt [hunk 1]
```diff
@@ -5,6 +5,54 @@ project (asgard
...
 )
 
+#-------------------------------------------------------------------------------
+#  Setup the kromult external project.
+#-------------------------------------------------------------------------------
+include (FetchContent)
+find_package (Git)
+
+#  Define a macro to register new projects.
+macro (register_project name dir url default_tag)
```
Comment (quantumsteve) [inline CMakeLists.txt]: Can (should?) this be written as a function with better control of scope?

File: CMakeLists.txt [hunk 2]
```diff
@@ -5,6 +5,54 @@ project (asgard
...
+            COMMAND ${GIT_EXECUTABLE} pull
+            WORKING_DIRECTORY ${${name}_SOURCE_DIR}
+        )
+    endif ()
+endmacro ()
+
+register_project (kronmult
+                  KRONMULT
+                  https://github.com/project-asgard/kronmult.git
+                  master
```
Comment (quantumsteve) [inline CMakeLists.txt]: Let's follow a specific sha instead of master. We don't want a change to kronmult to break asgard.

File: CMakeLists.txt [hunk 3]
```diff
@@ -5,6 +5,54 @@ project (asgard
...
+        GIT_TAG origin/${BUILD_TAG_${dir}}
+        SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/${dir}
+    )
+
+    FetchContent_MakeAvailable(${name})
+
+    if (GIT_FOUND)
+#  By default cmake clones projects in a headless state. After the repo is
+#  cloned checkout the requested tag so repo is in a working state.
+        execute_process(
```
Comment (quantumsteve) [inline CMakeLists.txt]: Why would a user (or a developer not modifying kronmult) want this?

File: CMakeLists.txt [hunk 4]
```diff
@@ -5,6 +5,54 @@ project (asgard
...
+            COMMAND ${GIT_EXECUTABLE} checkout ${BUILD_TAG_${dir}}
+            WORKING_DIRECTORY ${${name}_SOURCE_DIR}
+        )
+
+#  Add a taraget to pull the latest version before building. Note dependency is
+#  registered in the sub project CMakeList.txt. Not sure how this should handle
+#  multiple targets in a project yet. Name must match the subproject pull_
+#  dependency.
+        add_custom_target(
+            pull_${name}
```
Comment (quantumsteve) [inline CMakeLists.txt]: Changing the version of kronmult should require a PR.

File: contrib/contrib.cmake [hunk 5]
```diff
@@ -184,26 +184,26 @@ endif ()
 # link to Ed D'Azevedo's kronmult library, or download/build if not present
 #
 ###############################################################################
-if(ASGARD_USE_CUDA)
-    set(USE_GPU 1)
-endif()
+#if(ASGARD_USE_CUDA)
+#    set(USE_GPU 1)
```
Comment (quantumsteve) [inline contrib/contrib.cmake]: Where is `USE_GPU` being set?

File: CMakeLists.txt [hunk 6]
```diff
@@ -120,7 +168,7 @@ if(ASGARD_USE_CUDA)
   set_target_properties( kronmult_cuda PROPERTIES COMPILE_FLAGS "-arch sm_70 -g -lineinfo --ptxas-options=-O3")
   set_target_properties( kronmult_cuda PROPERTIES LINK_FLAGS "-Wl,-rpath,${CMAKE_BINARY_DIR}")
 endif()
-target_include_directories (kronmult_cuda PRIVATE ${CMAKE_SOURCE_DIR}/contrib/kronmult/ ${CMAKE_BINARY_DIR})
+#target_include_directories (kronmult_cuda PRIVATE ${CMAKE_SOURCE_DIR}/contrib/kronmult/ ${CMAKE_BINARY_DIR})
```
Comment (quantumsteve) [inline CMakeLists.txt]: please remove commented out code

File: contrib/contrib.cmake [hunk 7]
```diff
@@ -184,26 +184,26 @@ endif ()
...
+#endif()
 
-set(KRON_PATH "${CMAKE_CURRENT_BINARY_DIR}/contrib/kronmult/")
-set(KRON_INCLUDE_DIR "${CMAKE_SOURCE_DIR}/contrib/kronmult/")
+#set(KRON_PATH "${CMAKE_CURRENT_BINARY_DIR}/contrib/kronmult/")
+#set(KRON_INCLUDE_DIR "${CMAKE_SOURCE_DIR}/contrib/kronmult/")
```
Comment (quantumsteve) [inline contrib/contrib.cmake]: please remove commented out code

File: CMakeLists.txt [hunk 8]
```diff
@@ -5,6 +5,54 @@ project (asgard
...
+macro (register_project name dir url default_tag)
+    set (BUILD_TAG_${dir} ${default_tag} CACHE STRING "Name of the tag to checkout.")
+    set (BUILD_REPO_${dir} ${url} CACHE STRING "URL of the repo to clone.")
+
+#  Set up the sub project repository.
+    FetchContent_Declare(
+        ${name}
+        GIT_REPOSITORY ${BUILD_REPO_${dir}}
+        GIT_TAG origin/${BUILD_TAG_${dir}}
+        SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/${dir}
```
Comment (quantumsteve) [inline CMakeLists.txt]: I'd prefer the the SOURCE_DIR be inside `CMAKE_BUILD_DIR` so it doesn't pollute the source directory.

**Review by quantumsteve — CHANGES_REQUESTED** [review]
How do we remove the current git submodule from ASGarD?

Additional review comments attached

File: CMakeLists.txt (hunk 1 above)
Comment (cianciosa) [reply CMakeLists.txt]: I changed this to a function.

File: CMakeLists.txt (hunk 2 above)
Comment (cianciosa) [reply CMakeLists.txt]: Master has not been replaced with a specific git hash. I have also added a comment that changes to this hash should go through a pull request.

File: CMakeLists.txt (hunk 3 above)
Comment (cianciosa) [reply CMakeLists.txt]: This has been removed.

File: CMakeLists.txt (hunk 4 above)
Comment (cianciosa) [reply CMakeLists.txt]: Also removed.

File: contrib/contrib.cmake (hunk 5 above)
Comment (cianciosa) [reply contrib/contrib.cmake]: This is now set at https://github.com/cianciosa/asgard/blob/6becc11bcc1beed48c746da2a980bf70ee60eb5d/CMakeLists.txt#L154

File: CMakeLists.txt (hunk 6 above)
Comment (cianciosa) [reply CMakeLists.txt]: Removed

File: contrib/contrib.cmake (hunk 7 above)
Comment (cianciosa) [reply contrib/contrib.cmake]: Removed

File: CMakeLists.txt (hunk 8 above)
Comment (cianciosa) [reply CMakeLists.txt]: I have changed this to `${CMAKE_CURRENT_SOURCE_DIR}/contrib/${dir}` to match the location where the original git submodule was cloned to.

Comment (quantumsteve) [reply CMakeLists.txt]: Delete the existing submodule inside https://github.com/project-asgard/asgard/tree/develop/contrib

**Review by quantumsteve — APPROVED** [review]

**Pull request merged**